src/nanotron/models/llama.py

```diff
@@ -818,8 +818,10 @@ def get_block_compute_costs(self):
         d_qkv = model_config.hidden_size // model_config.num_attention_heads
         block_compute_costs = {
             # CausalSelfAttention (qkv proj + attn out) + MLP
-            LlamaDecoderLayer: 4 * model_config.num_attention_heads * d_qkv * model_config.hidden_size
-            + 3 * d_ff * model_config.hidden_size,
+            LlamaDecoderLayer: 2 * model_config.num_attention_heads * d_qkv * model_config.hidden_size # Q output projection
+            + 2 * model_config.num_key_value_heads * d_qkv * model_config.hidden_size # KV 
+            + 3 * d_ff * model_config.hidden_size # for the MLP (3 because of the gated mechanism)
+            + 2 * model_config.hidden_size, # for the layernorm
```

Comment:
I suggest removing the layer norm because its computational cost is relatively small compared to others. For matrix multiplication (MM), with dimensions m * k, k * n, the FLOPs are 2 * m * k * n.
```
LlamaDecoderLayer: 2 * model_config.num_attention_heads * d_qkv * model_config.hidden_size # Q projection
            + 4 * model_config.num_key_value_heads * d_qkv * model_config.hidden_size # KV projection
            + 2 * model_config.hidden_size ** 2 # Attention out projection
            + 6 * d_ff * model_config.hidden_size # for the MLP (3 because of the gated mechanism)
```

Reply:
agree with @zzhhjjj . Seems like other people as well remove the layernorm in the FLOPs computation (cf: https://github.com/EleutherAI/cookbook/blob/main/calc/calc_transformer_flops.py#L120)

Reply:
I can remove it but what's the cons of keeping it ? I mean that way we have a more exact FLOPs computation no?

Reply:
I'm not sure the flops of layer norm is 2 * model_config.hidden_size

Reply:
removed the LN, keeping the modification to take into account GQA, this one is not negligible.